http-mclient/src/Network/HTTP/MClient.hs

```diff
@@ -138,16 +140,57 @@ send' transactionDecoder decoder request = do
   sendRequest sendRequest' transactionDecoder decoder request
 
 sendRequest
-  :: SendRequest
+  :: forall env a . Env env
+  => SendRequest
   -> TransactionDecoder a
   -> ResponseDecoder a
   -> HTTP.Request
   -> MIO env (Result a)
 sendRequest sendRequest' transactionDecoder decoder request =
-  either transactionDecoder decoder
-    <$> Exception.tryJust selectException (liftIO $ sendRequest' request)
+  either transactionDecoder decoder <$> retriableSendRequest
   where
-    selectException :: HTTP.HttpException -> Maybe HTTP.HttpExceptionContent
-    selectException = \case
-      (HTTP.HttpExceptionRequest _request content) -> pure content
-      _other                                       -> empty
+    retriableSendRequest :: MIO env (Either HTTP.HttpExceptionContent (HTTP.Response LBS.ByteString))
+    retriableSendRequest
+      = Retry.retrying
+        limitedBackoff
+        isRetriable
+      $ const (Exception.tryJust selectException (liftIO $ sendRequest' request))
+      where
+        limitedBackoff :: Retry.RetryPolicy
+        limitedBackoff = Retry.exponentialBackoff 50_000 <> Retry.limitRetries (convertImpure retryLimit)
+
+        isRetriable
+          :: Retry.RetryStatus
+          -> Either HTTP.HttpExceptionContent (HTTP.Response LBS.ByteString)
+          -> MIO env Bool
+        isRetriable Retry.RetryStatus{..} = either process (const (pure False))
+          where
+            process :: HTTP.HttpExceptionContent -> MIO env Bool
+            process error
+              = Log.info
+                ( "Retrying failed request due to "
+                  <> showc error
+                  <> " attempt "
+                  <> showc rsIterNumber
+                  <> " of "
+                  <> showc retryLimit
+                )
```

Comment:
__question:__ Doesn't this presuppose that `isRequestNotSent error` is going to return true?

Maybe the result should be captured to a variable and the logging done conditionally?

Reply:
@dkubb good catch. Let me fix.